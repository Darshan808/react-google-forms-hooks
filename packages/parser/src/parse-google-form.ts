import fetch from "isomorphic-unfetch";
import { GoogleForm } from "@gforms-js/types";
import { extractRawFormData } from "./extract-raw-form-data";
import { parseRawFormData } from "./parse-raw-form-data";

const googleFormsHosts = ["docs.google.com", "forms.gle"];

const assertValidUrl = (formUrl: string): void => {
  const url = new URL(formUrl);

  if (!googleFormsHosts.includes(url.host)) {
    throw new Error(
      `Invalid Google Forms host. ${
        url.host
      } is expected to one of ${googleFormsHosts.join(", ")}`
    );
  }

  if (url.host === googleFormsHosts[0] && !url.pathname.endsWith("/viewform")) {
    throw new Error(`Please use the Google Forms public URL.`);
  }
};

const getFormHtml = async (formUrl: string) => {
  const response = await fetch(formUrl);
  const html = await response.text();
  return html;
};

interface ParseGoogleForm {
  (formUrl: string): Promise<GoogleForm>;
}

const parseGoogleForm: ParseGoogleForm = async (formUrl) => {
  assertValidUrl(formUrl);

  try {
    const html = await getFormHtml(formUrl);

    const rawFormData = extractRawFormData(html);

    return parseRawFormData(rawFormData);
  } catch (err) {
    throw new Error(`Failed to fetch Google Forms form. ${err}`);
  }
};

export { parseGoogleForm };
