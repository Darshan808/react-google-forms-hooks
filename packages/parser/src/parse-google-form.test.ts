import { test, expect, describe, beforeEach, vi } from "vitest";
import fs from "fs";
import path from "path";
import fetchMock from "fetch-mock";

import { parseGoogleForm } from "./parse-google-form";
import fetch from "node-fetch";

vi.mock("node-fetch", () => {
  return {
    default: fetchMock.sandbox(),
  };
});

const fetchMockSandbox = fetch as unknown as fetchMock.FetchMockSandbox;

const FORM_URL =
  "https://docs.google.com/forms/d/e/1FAIpQLSe5U3qvg8WHs4nkU-e6h2RlAD7fKoCkou6HO2w2-tXYIA_F8g/viewform";
const SHORTENED_FORM_URL = "https://forms.gle/vXSsfKSvpVJc3NMg9";

describe("parseGoogleForm", () => {
  beforeEach(() => {
    fetchMockSandbox.reset();
  });

  describe("form with all field types", () => {
    beforeEach(() => {
      const mockFormHtml = fs.readFileSync(
        path.join(__dirname, "..", "tests", "mocks", "form.html"),
        { encoding: "utf8" }
      );

      fetchMockSandbox.get(FORM_URL, mockFormHtml);
    });

    test("should parse the form correctly", async () => {
      const result = await parseGoogleForm(FORM_URL);
      expect(result).toMatchSnapshot();
    });
  });

  describe("when using a shortened url for the form", () => {
    beforeEach(() => {
      const mockFormHtml = fs.readFileSync(
        path.join(__dirname, "..", "tests", "mocks", "form.html"),
        { encoding: "utf8" }
      );

      fetchMockSandbox.get(SHORTENED_FORM_URL, mockFormHtml);
    });

    test("should parse the form correctly", async () => {
      const result = await parseGoogleForm(SHORTENED_FORM_URL);
      expect(result).toMatchSnapshot();
    });
  });

  describe("when the form has special characters", () => {
    beforeEach(() => {
      const mockFormHtml = fs.readFileSync(
        path.join(
          __dirname,
          "..",
          "tests",
          "mocks",
          "form-with-special-characters.html"
        ),
        { encoding: "utf8" }
      );

      fetchMockSandbox.get(FORM_URL, mockFormHtml);
    });

    test("should parse the form correctly", async () => {
      const result = await parseGoogleForm(FORM_URL);
      expect(result).toMatchSnapshot();
    });
  });

  describe("when the form is not from the correct host", () => {
    const INCORRECT_HOST_FORM = "https://example.com";

    test("throws an error", async () => {
      await expect(parseGoogleForm(INCORRECT_HOST_FORM)).rejects.toThrow();
    });
  });

  describe("when it fails to fetch the page", () => {
    beforeEach(() => {
      fetchMockSandbox.get(FORM_URL, {
        throws: new Error("Error"),
      });
    });

    test("throws an error", async () => {
      await expect(parseGoogleForm(FORM_URL)).rejects.toThrow();
    });
  });

  describe("when the form url doesnt correspond to the public url", () => {
    const PRIVATE_FORM_URL = FORM_URL.replace("/viewform", "");

    test("throws an error", async () => {
      await expect(parseGoogleForm(PRIVATE_FORM_URL)).rejects.toThrow();
    });
  });

  describe("when it can't find the script tag with the form data", () => {
    beforeEach(() => {
      const mockFormHtml = fs.readFileSync(
        path.join(__dirname, "..", "tests", "mocks", "invalid-form.html"),
        { encoding: "utf8" }
      );

      fetchMockSandbox.get(FORM_URL, mockFormHtml);
    });

    test("throws an error", async () => {
      await expect(parseGoogleForm(FORM_URL)).rejects.toThrow();
    });
  });

  describe("when it can't find the fbzx field", () => {
    beforeEach(() => {
      fetchMockSandbox.get(FORM_URL, "incorrect_html");
    });

    test("throws an error", async () => {
      await expect(parseGoogleForm(FORM_URL)).rejects.toThrow();
    });
  });

  describe("when parsing a form with an unsupported field", () => {
    beforeEach(() => {
      const mockFormHtml = fs.readFileSync(
        path.join(
          __dirname,
          "..",
          "tests",
          "mocks",
          "form-with-unsupported-field.html"
        ),
        { encoding: "utf8" }
      );

      fetchMockSandbox.get(FORM_URL, mockFormHtml);
    });

    test("should parse the form correctly", async () => {
      const result = await parseGoogleForm(FORM_URL);
      expect(result).toMatchSnapshot();
    });
  });
});
