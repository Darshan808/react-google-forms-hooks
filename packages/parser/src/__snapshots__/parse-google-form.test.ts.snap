// Vitest Snapshot v1

exports[`parseGoogleForm > form with all field types > should parse the form correctly 1`] = `
{
  "description": "form description",
  "fields": [
    {
      "id": "1973567836",
      "label": "Short answer question",
      "required": false,
      "type": "SHORT_ANSWER",
    },
    {
      "description": "With description",
      "id": "1027526081",
      "label": "Short answer question (required)",
      "required": true,
      "type": "SHORT_ANSWER",
    },
    {
      "id": "103082236",
      "label": "Paragraph question",
      "required": false,
      "type": "PARAGRAPH",
    },
    {
      "description": "With description",
      "id": "1047017538",
      "label": "Paragraph question (required)",
      "required": true,
      "type": "PARAGRAPH",
    },
    {
      "hasCustomOption": true,
      "id": "279371716",
      "label": "Multiple choice",
      "options": [
        {
          "label": "Option 1",
        },
        {
          "label": "Option 2",
        },
        {
          "label": "Option 3",
        },
      ],
      "required": false,
      "type": "MULTIPLE_CHOICE",
    },
    {
      "description": "With description",
      "hasCustomOption": false,
      "id": "37251313",
      "label": "Multiple choice (required)",
      "options": [
        {
          "label": "Option 1",
        },
        {
          "label": "Option 2",
        },
        {
          "label": "Option 3",
        },
      ],
      "required": true,
      "type": "MULTIPLE_CHOICE",
    },
    {
      "hasCustomOption": true,
      "id": "1523245781",
      "label": "Checkboxes",
      "options": [
        {
          "label": "Option 1",
        },
        {
          "label": "Option 2",
        },
        {
          "label": "Option 3",
        },
      ],
      "required": false,
      "type": "CHECKBOXES",
    },
    {
      "description": "With description",
      "hasCustomOption": false,
      "id": "357735393",
      "label": "Multiple choice (required)",
      "options": [
        {
          "label": "Option 1",
        },
        {
          "label": "Option 2",
        },
        {
          "label": "Option 3",
        },
      ],
      "required": true,
      "type": "CHECKBOXES",
    },
    {
      "id": "248358011",
      "label": "Dropdown",
      "options": [
        {
          "label": "Option 1",
        },
        {
          "label": "Option 2",
        },
        {
          "label": "Option 3",
        },
      ],
      "required": false,
      "type": "DROPDOWN",
    },
    {
      "description": "With description",
      "id": "882375941",
      "label": "Dropdown (required)",
      "options": [
        {
          "label": "Option 1",
        },
        {
          "label": "Option 2",
        },
        {
          "label": "Option 3",
        },
      ],
      "required": true,
      "type": "DROPDOWN",
    },
    {
      "id": "55542064",
      "label": "Linear scale",
      "maxNumberLabel": "",
      "minNumberLabel": "",
      "options": [
        {
          "label": "1",
        },
        {
          "label": "2",
        },
        {
          "label": "3",
        },
        {
          "label": "4",
        },
        {
          "label": "5",
        },
      ],
      "required": false,
      "type": "LINEAR_SCALE",
    },
    {
      "description": "With description",
      "id": "718227682",
      "label": "Linear scale (required)",
      "maxNumberLabel": "Max number label",
      "minNumberLabel": "Min number label",
      "options": [
        {
          "label": "0",
        },
        {
          "label": "1",
        },
        {
          "label": "2",
        },
        {
          "label": "3",
        },
        {
          "label": "4",
        },
        {
          "label": "5",
        },
        {
          "label": "6",
        },
        {
          "label": "7",
        },
        {
          "label": "8",
        },
        {
          "label": "9",
        },
        {
          "label": "10",
        },
      ],
      "required": true,
      "type": "LINEAR_SCALE",
    },
    {
      "columns": [
        {
          "label": "Column 1",
        },
        {
          "label": "Column 2",
        },
        {
          "label": "Column 3",
        },
      ],
      "id": "498463833",
      "label": "Multiple choice grid",
      "lines": [
        {
          "id": "498463833",
          "label": "Row 1",
        },
        {
          "id": "1786304611",
          "label": "Row 2",
        },
        {
          "id": "1040425630",
          "label": "Row 3",
        },
        {
          "id": "1772075609",
          "label": "Row 4",
        },
      ],
      "required": false,
      "type": "MULTIPLE_CHOICE_GRID",
    },
    {
      "columns": [
        {
          "label": "Column 1",
        },
        {
          "label": "Column 2",
        },
        {
          "label": "Column 3",
        },
      ],
      "description": "With description",
      "id": "1054835510",
      "label": "Multiple choice grid (required)",
      "lines": [
        {
          "id": "1054835510",
          "label": "Row 1",
        },
        {
          "id": "1839167326",
          "label": "Row 2",
        },
        {
          "id": "835156322",
          "label": "Row 3",
        },
        {
          "id": "335240212",
          "label": "Row 4",
        },
      ],
      "required": true,
      "type": "MULTIPLE_CHOICE_GRID",
    },
    {
      "columns": [
        {
          "label": "Column 1",
        },
        {
          "label": "Column 2",
        },
        {
          "label": "Column 3",
        },
        {
          "label": "Column 4",
        },
      ],
      "id": "657749623",
      "label": "Tick box grid",
      "lines": [
        {
          "id": "657749623",
          "label": "Row 1",
        },
        {
          "id": "924822411",
          "label": "Row 2",
        },
        {
          "id": "1950277088",
          "label": "Row 3",
        },
      ],
      "required": false,
      "type": "TICK_BOX_GRID",
    },
    {
      "columns": [
        {
          "label": "Column 1",
        },
        {
          "label": "Column 2",
        },
        {
          "label": "Column 3",
        },
        {
          "label": "Column 4",
        },
      ],
      "description": "With description",
      "id": "1456685102",
      "label": "Tick box grid (required)",
      "lines": [
        {
          "id": "1456685102",
          "label": "Row 1",
        },
        {
          "id": "261027955",
          "label": "Row 2",
        },
        {
          "id": "1093784376",
          "label": "Row 3",
        },
      ],
      "required": true,
      "type": "TICK_BOX_GRID",
    },
    {
      "id": "1162784826",
      "label": "Date",
      "required": false,
      "type": "DATE",
    },
    {
      "description": "With description",
      "id": "817206224",
      "label": "Date (required)",
      "required": true,
      "type": "DATE",
    },
  ],
  "fieldsPositionMap": {
    "1027526081": 1,
    "103082236": 2,
    "1047017538": 3,
    "1054835510": 13,
    "1162784826": 16,
    "1456685102": 15,
    "1523245781": 6,
    "1973567836": 0,
    "248358011": 8,
    "279371716": 4,
    "357735393": 7,
    "37251313": 5,
    "498463833": 12,
    "55542064": 10,
    "657749623": 14,
    "718227682": 11,
    "817206224": 17,
    "882375941": 9,
  },
  "formMetadata": {
    "action": "e/1FAIpQLSf5YK1mmRKPoFCFxNuH0XgyJ4bSRpFk3K5NkS-nlX4VBdysAw",
    "fbzx": "3047889860006435785",
    "fvv": 1,
    "pageHistory": 0,
  },
  "title": "Test form",
}
`;

exports[`parseGoogleForm > when parsing a form with an unsupported field > should parse the form correctly 1`] = `
{
  "description": "form description",
  "fields": [],
  "fieldsPositionMap": {},
  "formMetadata": {
    "action": "e/1FAIpQLScNatJq3N2INw_Dma5eSZEQyuMMf5nDCTZxwRgeztr3MLdmvg",
    "fbzx": "4517077496887845092",
    "fvv": 1,
    "pageHistory": 0,
  },
  "title": "Test form",
}
`;

exports[`parseGoogleForm > when the form has special characters > should parse the form correctly 1`] = `
{
  "fields": [
    {
      "hasCustomOption": true,
      "id": "1403784388",
      "label": "Hello, what's up? ",
      "options": [
        {
          "label": "Option 1",
        },
      ],
      "required": false,
      "type": "MULTIPLE_CHOICE",
    },
    {
      "hasCustomOption": false,
      "id": "162741379",
      "label": "Hello, what's up 2",
      "options": [
        {
          "label": "Option 1",
        },
      ],
      "required": false,
      "type": "MULTIPLE_CHOICE",
    },
  ],
  "fieldsPositionMap": {
    "1403784388": 0,
    "162741379": 1,
  },
  "formMetadata": {
    "action": "e/1FAIpQLSeT9rPzmpnaPRPrj6VXv9hXn-AYKYARfzstZmqIQJrkNYtxnQ",
    "fbzx": "6170924624432982469",
    "fvv": 1,
    "pageHistory": 0,
  },
  "title": null,
}
`;

exports[`parseGoogleForm > when using a shortened url for the form > should parse the form correctly 1`] = `
{
  "description": "form description",
  "fields": [
    {
      "id": "1973567836",
      "label": "Short answer question",
      "required": false,
      "type": "SHORT_ANSWER",
    },
    {
      "description": "With description",
      "id": "1027526081",
      "label": "Short answer question (required)",
      "required": true,
      "type": "SHORT_ANSWER",
    },
    {
      "id": "103082236",
      "label": "Paragraph question",
      "required": false,
      "type": "PARAGRAPH",
    },
    {
      "description": "With description",
      "id": "1047017538",
      "label": "Paragraph question (required)",
      "required": true,
      "type": "PARAGRAPH",
    },
    {
      "hasCustomOption": true,
      "id": "279371716",
      "label": "Multiple choice",
      "options": [
        {
          "label": "Option 1",
        },
        {
          "label": "Option 2",
        },
        {
          "label": "Option 3",
        },
      ],
      "required": false,
      "type": "MULTIPLE_CHOICE",
    },
    {
      "description": "With description",
      "hasCustomOption": false,
      "id": "37251313",
      "label": "Multiple choice (required)",
      "options": [
        {
          "label": "Option 1",
        },
        {
          "label": "Option 2",
        },
        {
          "label": "Option 3",
        },
      ],
      "required": true,
      "type": "MULTIPLE_CHOICE",
    },
    {
      "hasCustomOption": true,
      "id": "1523245781",
      "label": "Checkboxes",
      "options": [
        {
          "label": "Option 1",
        },
        {
          "label": "Option 2",
        },
        {
          "label": "Option 3",
        },
      ],
      "required": false,
      "type": "CHECKBOXES",
    },
    {
      "description": "With description",
      "hasCustomOption": false,
      "id": "357735393",
      "label": "Multiple choice (required)",
      "options": [
        {
          "label": "Option 1",
        },
        {
          "label": "Option 2",
        },
        {
          "label": "Option 3",
        },
      ],
      "required": true,
      "type": "CHECKBOXES",
    },
    {
      "id": "248358011",
      "label": "Dropdown",
      "options": [
        {
          "label": "Option 1",
        },
        {
          "label": "Option 2",
        },
        {
          "label": "Option 3",
        },
      ],
      "required": false,
      "type": "DROPDOWN",
    },
    {
      "description": "With description",
      "id": "882375941",
      "label": "Dropdown (required)",
      "options": [
        {
          "label": "Option 1",
        },
        {
          "label": "Option 2",
        },
        {
          "label": "Option 3",
        },
      ],
      "required": true,
      "type": "DROPDOWN",
    },
    {
      "id": "55542064",
      "label": "Linear scale",
      "maxNumberLabel": "",
      "minNumberLabel": "",
      "options": [
        {
          "label": "1",
        },
        {
          "label": "2",
        },
        {
          "label": "3",
        },
        {
          "label": "4",
        },
        {
          "label": "5",
        },
      ],
      "required": false,
      "type": "LINEAR_SCALE",
    },
    {
      "description": "With description",
      "id": "718227682",
      "label": "Linear scale (required)",
      "maxNumberLabel": "Max number label",
      "minNumberLabel": "Min number label",
      "options": [
        {
          "label": "0",
        },
        {
          "label": "1",
        },
        {
          "label": "2",
        },
        {
          "label": "3",
        },
        {
          "label": "4",
        },
        {
          "label": "5",
        },
        {
          "label": "6",
        },
        {
          "label": "7",
        },
        {
          "label": "8",
        },
        {
          "label": "9",
        },
        {
          "label": "10",
        },
      ],
      "required": true,
      "type": "LINEAR_SCALE",
    },
    {
      "columns": [
        {
          "label": "Column 1",
        },
        {
          "label": "Column 2",
        },
        {
          "label": "Column 3",
        },
      ],
      "id": "498463833",
      "label": "Multiple choice grid",
      "lines": [
        {
          "id": "498463833",
          "label": "Row 1",
        },
        {
          "id": "1786304611",
          "label": "Row 2",
        },
        {
          "id": "1040425630",
          "label": "Row 3",
        },
        {
          "id": "1772075609",
          "label": "Row 4",
        },
      ],
      "required": false,
      "type": "MULTIPLE_CHOICE_GRID",
    },
    {
      "columns": [
        {
          "label": "Column 1",
        },
        {
          "label": "Column 2",
        },
        {
          "label": "Column 3",
        },
      ],
      "description": "With description",
      "id": "1054835510",
      "label": "Multiple choice grid (required)",
      "lines": [
        {
          "id": "1054835510",
          "label": "Row 1",
        },
        {
          "id": "1839167326",
          "label": "Row 2",
        },
        {
          "id": "835156322",
          "label": "Row 3",
        },
        {
          "id": "335240212",
          "label": "Row 4",
        },
      ],
      "required": true,
      "type": "MULTIPLE_CHOICE_GRID",
    },
    {
      "columns": [
        {
          "label": "Column 1",
        },
        {
          "label": "Column 2",
        },
        {
          "label": "Column 3",
        },
        {
          "label": "Column 4",
        },
      ],
      "id": "657749623",
      "label": "Tick box grid",
      "lines": [
        {
          "id": "657749623",
          "label": "Row 1",
        },
        {
          "id": "924822411",
          "label": "Row 2",
        },
        {
          "id": "1950277088",
          "label": "Row 3",
        },
      ],
      "required": false,
      "type": "TICK_BOX_GRID",
    },
    {
      "columns": [
        {
          "label": "Column 1",
        },
        {
          "label": "Column 2",
        },
        {
          "label": "Column 3",
        },
        {
          "label": "Column 4",
        },
      ],
      "description": "With description",
      "id": "1456685102",
      "label": "Tick box grid (required)",
      "lines": [
        {
          "id": "1456685102",
          "label": "Row 1",
        },
        {
          "id": "261027955",
          "label": "Row 2",
        },
        {
          "id": "1093784376",
          "label": "Row 3",
        },
      ],
      "required": true,
      "type": "TICK_BOX_GRID",
    },
    {
      "id": "1162784826",
      "label": "Date",
      "required": false,
      "type": "DATE",
    },
    {
      "description": "With description",
      "id": "817206224",
      "label": "Date (required)",
      "required": true,
      "type": "DATE",
    },
  ],
  "fieldsPositionMap": {
    "1027526081": 1,
    "103082236": 2,
    "1047017538": 3,
    "1054835510": 13,
    "1162784826": 16,
    "1456685102": 15,
    "1523245781": 6,
    "1973567836": 0,
    "248358011": 8,
    "279371716": 4,
    "357735393": 7,
    "37251313": 5,
    "498463833": 12,
    "55542064": 10,
    "657749623": 14,
    "718227682": 11,
    "817206224": 17,
    "882375941": 9,
  },
  "formMetadata": {
    "action": "e/1FAIpQLSf5YK1mmRKPoFCFxNuH0XgyJ4bSRpFk3K5NkS-nlX4VBdysAw",
    "fbzx": "3047889860006435785",
    "fvv": 1,
    "pageHistory": 0,
  },
  "title": "Test form",
}
`;
